### PR TITLE
feat / Inventory skew single order pure mm

### PR DESCRIPTION
**Pull request by chowryan**
**Before submitting this PR, please make sure**:
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/", etc)

**Optional**:
- Related Github issue:
- Related Clubhouse Story:

**A description of the changes proposed in the pull request**:

### CODE REVIEW

**Comment by carlolm** [comment]
@chowryan 
puremm-v2 test failures:

```
test_check_if_active_orders_are_cancelled_every_tick (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ERROR
test_check_sufficient_balance (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... FAIL
test_confirm_active_bids_asks (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... FAIL
test_correct_price_correct_size (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ERROR
test_market_became_narrower (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ERROR
test_market_become_wider (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ERROR
test_multiple_orders_equal_sizes (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ok
test_multiple_orders_staggered_sizes (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ok
test_order_fills (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... FAIL
test_order_fills_after_cancellation (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... ERROR
test_strategy_after_user_cancels_orders (test_pure_market_making_v2.PureMarketMakingV2UnitTest) ... FAIL

======================================================================
ERROR: test_check_if_active_orders_are_cancelled_every_tick (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 266, in test_check_if_active_orders_are_cancelled_every_tick
    old_bid = self.strategy.active_bids[0][1]
IndexError: list index out of range
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_correct_price_correct_size (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 237, in test_correct_price_correct_size
    self.strategy.active_asks[0][1].price)
IndexError: list index out of range
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_market_became_narrower (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 325, in test_market_became_narrower
    bid_order: LimitOrder = self.strategy.active_bids[0][1]
IndexError: list index out of range
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_market_become_wider (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 302, in test_market_become_wider
    bid_order: LimitOrder = self.strategy.active_bids[0][1]
IndexError: list index out of range
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_order_fills_after_cancellation (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 347, in test_order_fills_after_cancellation
    bid_order: LimitOrder = self.strategy.active_bids[0][1]
IndexError: list index out of range
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_check_sufficient_balance (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 248, in test_check_sufficient_balance
    self.assertEqual(1, len(self.strategy.active_asks))
AssertionError: 1 != 0
    '1 != 0' = '%s != %s' % _common_shorten_repr(1, 0)
    '1 != 0' = self._formatMessage('1 != 0', '1 != 0')
>>  raise self.failureException('1 != 0')
    
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_confirm_active_bids_asks (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 231, in test_confirm_active_bids_asks
    self.assertEqual(1, len(self.strategy.active_bids))
AssertionError: 1 != 0
    '1 != 0' = '%s != %s' % _common_shorten_repr(1, 0)
    '1 != 0' = self._formatMessage('1 != 0', '1 != 0')
>>  raise self.failureException('1 != 0')
    
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_order_fills (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 279, in test_order_fills
    self.assertEqual(1, len(self.strategy.active_bids))
AssertionError: 1 != 0
    '1 != 0' = '%s != %s' % _common_shorten_repr(1, 0)
    '1 != 0' = self._formatMessage('1 != 0', '1 != 0')
>>  raise self.failureException('1 != 0')
    
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_strategy_after_user_cancels_orders (test_pure_market_making_v2.PureMarketMakingV2UnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_pure_market_making_v2.py", line 375, in test_strategy_after_user_cancels_orders
    self.assertEqual(1, len(self.strategy.active_bids))
AssertionError: 1 != 0
    '1 != 0' = '%s != %s' % _common_shorten_repr(1, 0)
    '1 != 0' = self._formatMessage('1 != 0', '1 != 0')
>>  raise self.failureException('1 != 0')
    
-------------------- >> begin captured logging << --------------------
hummingbot.strategy.pure_market_making.pure_market_making_v2: ERROR: Unknown error while generating order proposals.
Traceback (most recent call last):
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 257, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    orders_proposal = self.c_get_orders_proposal_for_market_info(
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 284, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_get_orders_proposal_for_market_info
    sizing_proposal = self._sizing_delegate.c_get_order_size_proposal(self,
  File "hummingbot/strategy/pure_market_making/constant_size_sizing_delegate.pyx", line 86, in hummingbot.strategy.pure_market_making.constant_size_sizing_delegate.ConstantSizeSizingDelegate.c_get_order_size_proposal
    (1.0 + float(buy_fees.percent)) *
TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
hummingbot.core.clock: ERROR: Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 145, in hummingbot.core.clock.Clock.backtest_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 265, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_tick
    self.c_execute_orders_proposal(market_info, orders_proposal)
  File "hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx", line 371, in hummingbot.strategy.pure_market_making.pure_market_making_v2.PureMarketMakingStrategyV2.c_execute_orders_proposal
    int64_t actions = orders_proposal.actions
AttributeError: 'NoneType' object has no attribute 'actions'
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 13 tests in 1.930s

FAILED (errors=5, failures=4)
```

**Comment by carlolm** [comment]
@chowryan need to add documentation to this, so we keep track of added features and so users understand what it's doing

**Comment by veecos2** [comment]
Also need unit test to test the specific inventory skew functionality.

**Comment by martinkou** [comment]
Same comment from me... we absolutely need unit test cases for the new functionality before this can be merged.

**Comment by chowryan** [comment]
@carlolm documentation will be separate PR

**Comment by chowryan** [comment]
@veecos2 and @martinkou yea im working on unit tests today

File: hummingbot/strategy/pure_market_making/inventory_skew_single_size_sizing_delegate.pyx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                quote_asset_balance += active_order.quantity * active_order.price
+            else:
+                has_active_ask = True
+                base_asset_balance += active_order.quantity
+
+        if has_active_bid and has_active_ask:
+            return SizingProposal([s_decimal_0], [s_decimal_0])
+
+        if self._inventory_target_base_percent is not None:
+            top_bid_price = Decimal(market.c_get_price(trading_pair, False))
```
Comment (martinkou) [inline hummingbot/strategy/pure_market_making/inventory_skew_single_size_sizing_delegate.pyx]: The scaling logic is pretty hard to read. I'd recommend adding at least two in-code documentation.

 1. The overall intention of this.
 2. Some rough explanation about the math. (esp. the `2 - ...` thing)

File: hummingbot/strategy/pure_market_making/inventory_skew_multiple_size_sizing_delegate.pyx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                quote_asset_balance += active_order.quantity * active_order.price
+            else:
+                has_active_ask = True
+                base_asset_balance += active_order.quantity
+
+        if has_active_bid and has_active_ask:
+            return SizingProposal([s_decimal_0], [s_decimal_0])
+
+        if self._inventory_target_base_percent is not None:
+            top_bid_price = Decimal(market.c_get_price(trading_pair, False))
```
Comment (martinkou) [inline hummingbot/strategy/pure_market_making/inventory_skew_multiple_size_sizing_delegate.pyx]: ditto

File: test/test_pure_market_making_v2.py [hunk 3]
```diff
@@ -516,3 +549,73 @@ def test_balance_for_multiple_staggered_orders(self):
...
+
+        maker_fill: OrderFilledEvent = self.maker_order_fill_logger.event_log[0]
+        self.assertEqual(TradeType.SELL, maker_fill.trade_type)
+        self.assertAlmostEqual(101, maker_fill.price)
+        self.assertAlmostEqual(Decimal("1.81818"), Decimal(str(maker_fill.amount)), places=4)
+
+        self.strategy.cancel_order(self.market_info, first_bid_order.client_order_id)
+        self.clock.backtest_til(self.start_timestamp + 2 * self.clock_tick_size + 1)
+        self.assertEqual(0, len(self.strategy.active_bids))
+        self.assertEqual(0, len(self.strategy.active_asks))
```
Comment (martinkou) [inline test/test_pure_market_making_v2.py]: The unit tests here are not adequate... the important thing to test is whether the orders created have the scaled size in an unbalanced situation (after that 5.0 order fill). But instead you're just testing for the order fill events here.

File: test/test_pure_market_making_v2.py [hunk 4]
```diff
@@ -516,3 +549,73 @@ def test_balance_for_multiple_staggered_orders(self):
...
+
+    def test_inventory_skew_multiple_order_strategy(self):
+        self.clock.remove_iterator(self.strategy)
```
Comment (martinkou) [inline test/test_pure_market_making_v2.py]: Same for this unit test, the order sizes are not being tested.

**Comment by mifeng** [comment]
FYI @chowryan I'm going to push a commit that makes a small mod to config walkthrough text.

**Pull request merged**